### DISCUSSION TITLE
图层系统

### DISCUSSION THREAD

**Discussion by fifth-light** [TouchController/TouchController#123] [想法]
目前在做新配置界面的开发，打算引入一个图层系统，以解决各种问题：

- #97
- #86
- #80
- #75

我目前的想法是：可以创建任意个图层，这些图层会按用户指定的顺序叠加在一起，每个图层都可以有自己的控制组件，并且可以设置每个图层的显示条件（游泳中、飞行中、潜行中、疾跑中、在船上、在矿车上、在猪上、在载具上等等）。

对于这个图层系统，不知道各位有什么想法？

**Comment by fengzhou0w0** [comment]
> 这些图层会按用户指定的顺序叠加在一起，每个图层都可以有自己的控制组件，并且可以设置每个图层的显示条件（游泳中、飞行中、潜行中、疾跑中、在船上、在矿车上、在猪上、在载具上等等）。

想问一下，这个图层系统的最终开发目标，会到达以下哪一步？

①：目标为还原基岩版的各种状态下的ui界面的，基础版图层系统。图层数量以及命名是定死的，比如主图层、飞行图层、驾船图层等，玩家只能拥有修改删减图层内各个按钮的自由度。

②：目标为实现完整版的，可完全自定义的图层系统。除了要开发①中那些需要固定的，游戏必要的基础图层，还可允许玩家自行增减自定义图层，以及提供设计自定义图层之间交互逻辑的功能。（参考fcl按键，这样的话就是个大工程了，可能为了实现这个，还需要先增加自定义按键等功能吧）

**Reply by fifth-light** [reply]
实现你说的完整版，但是没有所谓基础图层，每个图层都是可自定义的。自定义按键暂且不实现。反正是个大工程，顺带把多游戏版本支持一起做了

**Reply by fengzhou0w0** [reply]
> 但是没有所谓基础图层，每个图层都是可自定义的。

我明白了，就是说每一组按键预设（假设已实现#56的功能），我们都会先给一些图层，把主图层潜行图层划船图层等先准备好，初始状态和原基岩版ui对应好了，再让玩家按自己的需要自行增减修改，是这个思路吧。

我觉得有必要增加一个预设初始化，万一玩家不小心给图层删改地不能正常玩了，这时候的玩家需要一个从头开始调整图层及按键的机会。

**Reply by fifth-light** [reply]
是这个思路。预设功能我打算跟着图层功能一起加入了，然后会有所有图层的预设和单个图层的预设，这样就可以恢复单个图层了。

**Reply by fifth-light** [reply]
> 以及提供设计自定义图层之间交互逻辑的功能

这点可能不会有，切换图层的方式我打算设计成只能是根据玩家不同的状态切换，不会有其他的切换方式。

**Reply by fengzhou0w0** [reply]
> 顺带把多游戏版本支持一起做了

好像issue里边还有些bug没修完，搞完多版本再修，修起来会不会很麻烦，每次更新都要发一大堆不同版本的文件

**Reply by fengzhou0w0** [reply]
> 这点可能不会有，切换图层的方式我打算设计成只能是根据玩家不同的状态切换，不会有其他的切换方式。

合理，不过对于“根据玩家状态切换图层”的话，可以多给一些“玩家状态”选择。比如，可以预留一个打开背包的状态。假如未来增加了shift、r、右键等整理背包的按键，就可以在这个预留的图层中使用了。

**Reply by fengzhou0w0** [reply]
我记得以前有issue提出，使用触摸屏win平板的玩家刚需一个随时能点击的esc键。这个需求也会在图层系统中解决么？还是说到时候靠其他方式实现，比如悬浮球、贴边条之类的。

**Reply by fifth-light** [reply]
> 搞完多版本再修，修起来会不会很麻烦

不会的，这点没关系，我的思路是把核心逻辑部分抽离为一个模块，然后每个版本就是一个子模块。其实每个版本自己的代码不多，这样大部分 bug 都会发生在主模块上。至于每一次发布都会包含所有的游戏版本这个事，我可能会搞自动发布来解决。

> 可以预留一个打开背包的状态

这个之后考虑，目前只考虑不打开游戏内界面的情况。要做这个功能的话那就的话不一定是背包了，应该是打开任何游戏内界面都算一个状态。

> 使用触摸屏win平板的玩家刚需一个随时能点击的esc键

目前不打算做在图层系统里。我打算就在屏幕底端或者顶端绘制一个关闭按钮可以了。

**Comment by fengzhou0w0** [comment]
如果未来要实现f1隐藏所有键位绘制的功能，到时候的设计是直接隐藏所有图层的按键对吧，图层之间的切换，交互应该不受影响吧。

**Reply by fifth-light** [reply]
直接隐藏所有图层，也就是不进行所有图层的绘制，其他不受影响

**Comment by WhiteElephant-abc** [comment]
> 目前在做新配置界面的开发，打算引入一个图层系统，以解决各种问题：
> 
> * [游泳时下蹲按键没有隐藏 #97](https://github.com/fifth-light/TouchController/issues/97)
> * [交通工具相关键位适配 #86](https://github.com/fifth-light/TouchController/issues/86)
> * [远古版创造模式飞行控制 #80](https://github.com/fifth-light/TouchController/issues/80)
> * [飞行时下蹲按钮不会隐藏 #75](https://github.com/fifth-light/TouchController/issues/75)
> 
> 我目前的想法是：可以创建任意个图层，这些图层会按用户指定的顺序叠加在一起，每个图层都可以有自己的控制组件，并且可以设置每个图层的显示条件（游泳中、飞行中、潜行中、疾跑中、在船上、在矿车上、在猪上、在载具上等等）。
> 
> 对于这个图层系统，不知道各位有什么想法？

还有 #40

**Comment by WhiteElephant-abc** [comment]
> 每个图层都可以有自己的控制组件，并且可以设置每个图层的显示条件

就是说不同场景显示不同控件是吗

> 可以设置每个图层的显示条件（游泳中、飞行中、潜行中、疾跑中、在船上、在矿车上、在猪上、在载具上等等）

那么这样是准备添加自定义按键的功能吗？毕竟不是所有控价都可以在这些状态下使用

> 可以创建任意个图层，这些图层会按用户指定的顺序叠加在一起

不然的话我是想不到什么按键是需要在一些条件下隐藏的，按基岩版高版本的逻辑
但是如果按基岩版低版本的话，确实有这个必要，不过这样的话就会出现一些问题，比如有些按键功能相同，例如上升、上浮、跳跃

> 这个之后考虑，目前只考虑不打开游戏内界面的情况。要做这个功能的话那就的话不一定是背包了，应该是打开任何游戏内界面都算一个状态。

我觉得打开界面才是这个图层系统最大的应用

> 目前在做新配置界面的开发，打算引入一个图层系统，以解决各种问题：
> 
> * [游泳时下蹲按键没有隐藏 #97](https://github.com/fifth-light/TouchController/issues/97)
> * [交通工具相关键位适配 #86](https://github.com/fifth-light/TouchController/issues/86)
> * [远古版创造模式飞行控制 #80](https://github.com/fifth-light/TouchController/issues/80)
> * [飞行时下蹲按钮不会隐藏 #75](https://github.com/fifth-light/TouchController/issues/75)

如果只是解决这类问题的话，可以不做自定义功能，毕竟例如 #86 ，这些都是预制的，如果（以划船UI为例）两个相同贴图的控件一个，功能不同......这就无法通过自定义复现这一图层，顶多调整一下按键位置、大小，如果说再把这个图层改为其他触发逻辑，也不太合理，毕竟划船UI的逻辑是很特殊的，不能再其他场景下适用

**Reply by fengzhou0w0** [reply]
>比如有些按键功能相同，例如上升、上浮、跳跃

我觉得这个问题好解决，只要切换状态的时候，把上一图层的某个按钮隐藏的同时，下一图层相同位置与大小的按钮再显示出来，就能做出这种效果了。就是可能需要添加一个按键设置绑定的功能。

**Reply by fifth-light** [reply]
> 就是说不同场景显示不同控件是吗

是的，根据玩家不同状态显示不同的控件。

> 那么这样是准备添加自定义按键的功能吗？

目前并不准备添加。之后会添加，但是和目前图层系统的设计关系不大了。

> 不然的话我是想不到什么按键是需要在一些条件下隐藏的

在老版本的触控下，部分条件下一些按钮会隐藏的。还有载具相关的适配也要考虑。

> 有些按键功能相同，例如上升、上浮、跳跃

这点没关系，只要不同图层的重叠在一起，然后保证同一时刻只出现一个图层就可以。我目前的事件处理逻辑可以正确处理这种情况。

> 我觉得打开界面才是这个图层系统最大的应用

目前不打算实现，日后考虑。我的核心目的还是实现老版本的基岩版操作。

> 如果只是解决这类问题的话，可以不做自定义功能

目前没有图层系统的情况就是这样的。我是因为不想在代码里写死控制布局才考虑做图层系统的。如果要写死每个组件的位置、大小的话我连控制布局自定义功能都不会写出来，或许就做成现在基岩版的三个选项算了。

**Reply by WhiteElephant-abc** [reply]
> > 有些按键功能相同，例如上升、上浮、跳跃
> 
> 这点没关系，只要不同图层的重叠在一起，然后保证同一时刻只出现一个图层就可以。我目前的事件处理逻辑可以正确处理这种情况。

我的意思是如果自定义，是不是可以在同一图层内添加这些按钮，而玩家还要区分这些按钮并添加到正确的图层中

**Reply by fifth-light** [reply]
> 可以在同一图层内添加这些按钮，而玩家还要区分这些按钮并添加到正确的图层中

我倒觉得可以实现图层内复制粘贴组件，并且添加在多个图层内移动相同组件的功能来解决这个问题。说到底其实常用图层也就三个、一个是常驻图层、一个是飞行或者游泳的图层、还有一个是正常状态下的图层，别的图层都是用于载具相关的场景的。
